pkg/hooks/webhook.go

```diff
@@ -99,15 +146,32 @@ func (w *WebhookExecutor) Call(request interface{}, response interface{}) error
 		logging.Logger.V(6).Info("Webhook response", "type", w.hookType, "url", w.url, "body", rawResponse)
 	}
 
-	// Check status code.
-	if resp.StatusCode != http.StatusOK {
+	if cacheEntry != nil {
+		// According to https://datatracker.ietf.org/doc/html/rfc7232
+		// When 'If-None-Match' is present, backend should respond 304 when object has not changed
+		if resp.StatusCode == 304 {
+			// TODO: Find a way to deep copy from cacheEntry.Response to response and switch cacheEntry.Response to store decoded response
+			if err = k8sjson.Unmarshal(cacheEntry.Response, response); err != nil {
+				return fmt.Errorf("can't unmarshal response: %w", err)
+			}
+			return nil
+		}
+	} else if resp.StatusCode != http.StatusOK {
 		return fmt.Errorf("remote error: %s", respBody)
 	}
 
 	// Decode response.
-	if err := k8sjson.Unmarshal(respBody, response); err != nil {
+	if err = k8sjson.Unmarshal(respBody, response); err != nil {
 		return fmt.Errorf("can't unmarshal response: %w", err)
 	}
+
+	if cacheEntry != nil {
```

Comment:
suggest: This condition seems repeated & hence is prone to bugs & readability issues.
Can we try following:
```go
if cacheEntry == nil {
  k8sjson.Unmarshal(respBody, response);
}

// check for status code(s) here
```

Reply:
Please check if we can keep the condition for `resp.StatusCode != http.StatusOK` as-is & then add this new logic.
E.g.
```go
if resp.StatusCode != http.StatusOK {
  ...
}

if cacheEntry == nil {
  ...
}

// move the logic for status code 304 here
// logic to extract eTag
```

Reply:
> suggest: This condition seems repeated & hence is prone to bugs & readability issues. Can we try following:
> 
> ```go
> if cacheEntry == nil {
>   return k8sjson.Unmarshal(respBody, response);
> }
> 
> // move the logic for status code(s) here
> // logic to extract eTag
> ```

Nope, we do not want to cache broken json, so we store it after it is parsed

Reply:
> Please check if we can keep the condition for `resp.StatusCode != http.StatusOK` as-is & then add this new logic. E.g.
> 
> ```go
> if resp.StatusCode != http.StatusOK {
>   ...
> }
> 
> if cacheEntry == nil {
>   ...
> }
> 
> // move the logic for status code 304 here
> // logic to extract eTag
> ```

Nope, we can't, reponse status 304 will fail `resp.StatusCode != http.StatusOK` condition.
And it is valid only when etag is enabled.

Reply:
in general, agree, it looks like spagetti-code, there is only one way to make it better is to split code in multiple methods and either call them on condition or have two different  Call methods one for when etag is enabled, another - disabled.
Wdt ? do you want me to go this route ?

Reply:
ok. Lets keep your changes as-is.